ITE/6/README.md

```diff
@@ -0,0 +1,743 @@
+
+reproducible_build(attestation):
+  attestation is signed by 'ReproducibleBuilds'
+  attestation.attestation_type == 'https://example.com/ReproducibleBuild/v1'
+
+verifiable_build(attestation):
+  return (hermetic_github_action(attestation) or
+          hermetic_cloud_build(attestation) or
+          hermetic_codebuild(attestation))
+
+hermetic_github_action(attestation):
+  attestation is signed by 'GitHubActions'
+  attestation.attestation_type == 'https://example.com/GitHubActionProduct/v1'
+  attestation.details.hermetic == true
+
+hermetic_cloud_build(attestation):
+  attestation is signed by 'GoogleCloudBuild'
+  attestation.attestation_type == 'https://example.com/GoogleCloudBuildProduct/v1'
+  attestation.details.no_network == true
+
+hermetic_cloud_build(attestation):
+  attestation is signed by 'AwsCodeBuild'
+  attestation.attestation_type == 'https://example.com/AwsCodeBuildProduct/v1'
+  attestation.details.no_network == true
+
+# Types of artifact IDs considered by `lookup attestations for <X>`.
+allowed_artifact_id_types = [
+  'sha256', 'sha512', 'container_image_digest', 'git_commit',
+]
+```
+
+## Attestations
+
+Let's take a look at one example team's software supply chain.
+
+![drawing](attestation_supply_chain.png)
+
+*   Top-level code repository is "https://github.com/my-company/my-product".
+    *   This defines submodules and the GitHub Actions workflows.
+*   Vulnerability scan is provided by an in-house scanner.
+*   Docker image is produced by the GitHub Actions "Build" workflow.
+    *   In the hypothetical "hermetic" mode, this records all dependent
+        submodules and build tools.
+
+This corresponds to the following attestations. Assume each is signed by the
+appropriate party; we only show the claim here.
+
+```json
+{
+  "attestation_type": "https://example.com/CodeReview/v1",
+  "subject": { "git_commit": "859b387b985ea0f414e4e8099c9f874acb217b94" },
+  "details": {
+    "timestamp": "2020-04-12T13:50:00Z",
+    "repo_type": "git",
+    "repo_url": "https://github.com/my-company/my-product",
+    "repo_branch": "master",
+    "code_reviewed": true
+  }
+}
+```
+
+```json
+{
+  "attestation_type": "https://example.com/CodeReview/v1",
+  "subject": { "git_commit": "2f02c094e6a9afe8e889c3f1d3cb66b437797af4" },
+  "details": {
+    "timestamp": "2020-04-12T13:50:00Z",
+    "repo_type": "git",
+    "repo_url": "https://github.com/my-company/submodule1",
+    "repo_branch": "master",
+    "code_reviewed": true
+  }
+}
+```
+
+```json
+{
+  "attestation_type": "https://example.com/CodeReview/v1",
+  "subject": { "git_commit": "5215a97a7978d8ee0de859ccac1bbfd2475bfe92" },
+  "details": {
+    "timestamp": "2020-04-12T13:50:00Z",
+    "repo_type": "git",
+    "repo_url": "https://github.com/my-company/submodule2",
+    "repo_branch": "master",
+    "code_reviewed": true
+  }
+}
+```
+
+```json
+{
+  "attestation_type": "https://example.com/VulnerabilityScan/v1",
```

Comment:
As an aside, I worry about this sort of attestations (e.g. vulnerability scans), since they seem to break monotonicity. Is this something that we should worry about?

Reply:
Yes, that is a concern in general, but here it is monotonic. On line 415, we check for a positive assertion that there are zero vulnerabilities. Once we define that schema for real, we will be sure to make this point clear to policy authors. We also plan to provide policy "templates" that people instantiate, and we'll design those templates to be monotonic.

That said, the situation is a bit more nuanced. There are actually two use cases, only the first of which requires monotonicity:
- Admission control: only allow "good" things. In this case, we require positive attestation of goodness, and allow even if there exists an attestation of badness.
- Remediation: prevent "bad" things on a best effort basis. In this case, we would want to know about the attestation of badness, even if the artifact was previously allowed.

A policy that encompasses both might be "allow if there exists an attestation showing zero vulnerabilities, but deny if there exists an attestation showing a high severity vulnerability."

@nenaddedic FYI